### PR TITLE
Add vary header on all requests when plone.rest is installed.

### LINKED ISSUE #73: Content negotiation leads to cache poisoning
As explained on [a post in the community forum](https://community.plone.org/t/6658?u=hvelarde), intermediate proxies cause issues if we don't include a `Vary: Accept` header.

AFAIK, we can fix this using two approaches:

* adding the `Vary` header to all `GET` and `HEAD` responses
* document how to deal with this issue using plone.app.caching

IMO, the later is harder and prone to errors.

### PR DESCRIPTION
**Pull request by erral**
This fixes #73

### CODE REVIEW

File: src/plone/rest/demo.py [hunk 1]
```diff
@@ -99,3 +99,14 @@ class NamedOptions(BaseService):
 
     def data(self):
         return {'service': 'named options'}
+
+
+class MultiVaryService(Service):
+
+    def render(self):
+        self.request.response.setHeader("Content-Type", "application/json")
```
Comment (hvelarde) [inline src/plone/rest/demo.py]: use single quotes instead.

File: src/plone/rest/vary.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# -*- coding: utf-8 -*-
+def add_vary_accept_header(event):
+    """Add the ``Vary: Accept`` header for all responses
```
Comment (hvelarde) [inline src/plone/rest/vary.py]: single line docstring.

File: README.rst [hunk 3]
```diff
@@ -149,6 +149,20 @@ OPTIONS::
   $ http --auth admin:admin OPTIONS localhost:8080/Plone/doc1 Accept:application/json
 
 
+Content Negotiation and caching
+-------------------------------
+
+Due to the dual behavior of URLs when using Content Negotiation (the same URL can provide
```
Comment (hvelarde) [inline README.rst]: use semantic line breaks.

**Comment by hvelarde** [comment]
anything else missing?

**Comment by jensens** [comment]
Does it work with Plone 5.2 and the WSGI publisher used in Python 3?

**Comment by hvelarde** [comment]
I don't know if this work Plone 5.2 and the WSGI publisher used in Python 3, but this is a bug fix for a serious problem that can be even used as a vector attack on sites that have a caching proxy. please review and merge it if you have no further comments and let's have a new release ASAP. we can fix the earlier later.

**Comment by tisto** [comment]
Run PR tests agains Plone 5.2: https://jenkins.plone.org/job/pull-request-5.2/1060/console

**Comment by hvelarde** [comment]
@tisto awesome, everything is green now.

**Comment by hvelarde** [comment]
hi guys, any news on this?

**Comment by jensens** [comment]
I am pretty sure this needs discussion. After quickly reading into the topic it seems we will reduce efficiency of caching. Every browser sends different default vary headers, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation/List_of_default_Accept_values
I agree, we need a solution here. Is this the solution or is there a different, more efficient way?
cc @plone/framework-team

**Comment by hvelarde** [comment]
@jensens you're totally right and seems we're not the only one facing this challenge: [Drupal guys implemented a parameter](https://books.google.com.br/books?id=GUXiCwAAQBAJ&lpg=PA301&ots=HPbIytCSKZ&dq=%22vary%3A%20accept%22%20varnish%20-encoding%20-language%20_format&hl=pt-BR&pg=PA301#v=onepage&q=%22vary:%20accept%22%20varnish%20-encoding%20-language%20_format&f=false) instead of using the `Vary: Accept` header.

I don't like that solution; I think we must keep the header and document how to workaround the issue in Varnish configuration: normalizing the `Accept` header in `vcl_recv` could be an option like described here:

https://varnish-cache.org/docs/trunk/users-guide/increasing-your-hitrate.html#http-vary

but, yes, we need more discussion on this.

**Comment by jensens** [comment]
also interesting (we are not the only ones with the problem) https://info.varnish-software.com/blog/introducing-vmod-accept

**Comment by hvelarde** [comment]
@jensens could you please take this issue to the next @plone/framework-team meeting? I know August is a hard month as everybody seems to be out of the office on the Northern hemisphere, but we need to find a solution ASAP.

**Comment by hvelarde** [comment]
@plone/framework-team any news on this issue?

**Comment by tisto** [comment]
@hvelarde we discussed this at the last fwt meeting. Bottom line is somebody needs to invest time to investigate this. It would help a lot if you could start writing documentation of how the system is supposed to behave and how varnish would need to be configured. I will try to find time to look into this but I can not promise anything right now...

**Comment by hvelarde** [comment]
@tisto that's already documented in #73; resuming: if I have a Plone instance with plone.restapi installed I should be able get two different responses if I ask from some piece of content as `text/html` or as `application/json`; without the `Vary` header intermediate proxies will store only one copy of the response resulting in cache poisoning which can be used as a vector for a site attack.

**Comment by tisto** [comment]
@hvelarde I am aware of that. As @jensens already mentioned, this is not good enough to merge right now. The matter is more complex and requires a more in-depth analysis and investigation. If you are not willing to do that, you have to wait until somebody else has time for it...

**Comment by hvelarde** [comment]
@tisto just to be clear, if I knew how to fix this I wouldn't be here asking for help.

**Comment by mamico** [comment]
Working with plone.rest/restapi/volto, I have encountered problems discussed here.

My two cents:

1. The '/api' prefix option has some disadvantages: not intercepting different content negotiations managed by plone.rest (eg. someone could implement both application/json and and application/xml), it could be difficult to manage when using complex vhm rules and "/api" is not near the root path;
2. extension `some/path.json` or custom view `some/path/@@api.json` or adding another custom header, AFAIK doesn't like to many people here ...
3. Custom varnish rules may be too complicated and not easy to explain to people using a different cache server

Eventually "Vary: Accept" might be the best option. 
If the Vary header will be added only for the API request, then here https://github.com/plone/plone.rest/blob/master/src/plone/rest/events.py#L25 
```
    if service_id is not None:
        alsoProvides(request, IAPIRequest)
 +      request.response.appendHeader('Vary'. 'Accept')
        request._rest_service_id = service_id
```
and not for all requests, there is no risk of reducing the effectiveness of the cache (as correctly highlighted by @jensens).

**Comment by jensens** [comment]
With the new `++api++` traverser this is no longer needed. I close this. Please reopen if you think it is needed anyway.

**Pull request closed without merging**